sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveAggsThroughLeftSemiAntiJoin.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.plans.{LeftAnti, LeftSemi}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Join, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.{AGGREGATE, LEFT_SEMI_OR_ANTI_JOIN}
+
+/**
+ * Remove the aggregation from left semi/anti join if the same aggregation has already been done
+ * on left side.
+ */
+object RemoveAggsThroughLeftSemiAntiJoin extends Rule[LogicalPlan] {
+  // Transform down to remove more Aggregates.
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformDownWithPruning(
+    _.containsAllPatterns(AGGREGATE, LEFT_SEMI_OR_ANTI_JOIN), ruleId) {
+    case agg @ Aggregate(grouping, aggExps, j @ Join(left: Aggregate, _, LeftSemi | LeftAnti, _, _))
+      if agg.groupOnly && left.groupOnly &&
+        aggExps.forall(e => left.aggregateExpressions.exists(_.semanticEquals(e))) &&
```

Comment:
This seems to assume `RemoveRepetitionFromGroupExpressions` always, but the test suite doesn't cover that.
Can we add a more complex test case which needs `RemoveRepetitionFromGroupExpressions` explicitly?

Reply:
Do you mean this condition?
```scala
aggExps.forall(e => left.aggregateExpressions.exists(_.semanticEquals(e)))
```

Reply:
Yes~

Reply:
Added it.
https://github.com/apache/spark/blob/d89da31d57c6c52d28c35d175cd219cf6f1f13cd/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregatesInLeftSemiAntiJoinSuite.scala#L91-L100